ssg/rule_dir_stats.py

```diff
@@ -0,0 +1,462 @@
+    right_only_rule_ids = right_rule_ids.difference(left_rule_ids)
+    common_rule_ids = left_rule_ids.intersection(right_rule_ids)
+
+    left_restricted = subset_dict(left_rules, left_only_rule_ids)
+    left_common = subset_dict(left_rules, common_rule_ids)
+    right_restricted = subset_dict(right_rules, right_only_rule_ids)
+    right_common = subset_dict(right_rules, common_rule_ids)
+
+    left_only_data = walk_rules(args, left_restricted, oval_func, remediation_func)
+    right_only_data = walk_rules(args, right_restricted, oval_func, remediation_func)
+    l_c_d, r_c_d, c_d = walk_rules_parallel(args, left_common, right_common,
+                                            oval_func, remediation_func)
+
+    left_changed_data = l_c_d
+    right_changed_data = r_c_d
+    common_data = c_d
+
+    return (left_only_data, right_only_data, left_changed_data, right_changed_data, common_data)
+
+
+def walk_rules_diff_stats(results):
+    """
+    Takes the results of walk_rules_diff (results) and generates five sets of
+    output statistics: left_only rules output, right_only rules output,
+    shared left output, shared right output, and shared common output, as a
+    five-tuple, where each tuple element is equivalent to walk_rules_stats on
+    the appropriate set of rules.
+
+    Can assert.
+    """
+
+    assert len(results) == 5
+
+    output_data = []
+
+    for data in results:
+        affected_rules, verbose_output = data
+
+        affected_ovals = 0
+        affected_remediations = 0
+        all_affected_remediations = 0
+        affected_remediations_type = defaultdict(lambda: 0)
+        all_output = []
+
+        for rule_id in sorted(verbose_output):
+            rule_output = verbose_output[rule_id]
+            _results = walk_rule_stats(rule_output)
+
+            affected_ovals += _results[0]
+            affected_remediations += _results[1]
+            all_affected_remediations += _results[2]
+            for key in _results[3]:
+                affected_remediations_type[key] += _results[3][key]
+
+            all_output.extend(_results[4])
+
+        output_data.append((affected_rules, affected_ovals,
+                            affected_remediations, all_affected_remediations,
+                            affected_remediations_type, all_output))
+
+    assert len(output_data) == 5
+
+    return tuple(output_data)
+
+
+def missing_oval(rule_obj):
+    """
+    For a rule object, check if it is missing an oval.
+    """
+
+    rule_id = rule_obj['id']
+    check = len(rule_obj['ovals']) > 0
+    if not check:
+        return "\trule_id:%s is missing all OVALs" % rule_id
+
+
+def missing_remediation(rule_obj, r_type):
+    """
+    For a rule object, check if it is missing a remediation of type r_type.
+    """
+
+    rule_id = rule_obj['id']
+    check = len(rule_obj['remediations'][r_type]) > 0
+    if not check:
+        return "\trule_id:%s is missing %s remediations" % (rule_id, r_type)
```

Comment:
Nitpick: I know it is part of providing this generic framework, but is a bit annoying that it outputs a line for each type of remediation:
```
	rule_id:accounts_password_pam_lcredit is missing anaconda remediations
	rule_id:accounts_password_pam_lcredit is missing ansible remediations
	rule_id:accounts_password_pam_lcredit is missing bash remediations
	rule_id:accounts_password_pam_lcredit is missing puppet remediations
```